ravenml/data/write_dataset.py

```diff
@@ -0,0 +1,166 @@
+import os, shutil, time, json
+import contextlib2
+from pathlib import Path
+from object_detection.dataset_tools import tf_record_creation_util
+from random import shuffle
+from datetime import datetime
+from ravenml.data.interfaces import CreateInput
+from ravenml.utils.question import cli_spinner, cli_spinner_wrapper
+
+class DatasetWriter(object):
+
+    def __init__(self, create: CreateInput, **kwargs):
+        self.associated_files = kwargs['associated_files']
+        self.related_data_prefixes = kwargs['related_data_prefixes']
+        
+        metadata = create.metadata
+        self.num_folds = metadata['kfolds'] if metadata.get('kfolds') else 5 # Never actually used
+        self.test_percent = metadata['test_percent'] if metadata.get('test_percent') else .2
+        self.label_to_int_dict = {}
+        self.dataset_path = create.dataset_path
+        self.dataset_name = metadata['dataset_name']
+        self.cache = create.plugin_cache
+    
+    def construct_all(self, *args, **kwargs):
```

Comment:
construct_all, export_as_TFExample, and write_additional_files are expected to be overloaded by the plugins if they intend to use them